### ISSUE TITLE
printing prepended data may cause errant �

### ISSUE THREAD

**Issue by jtmoon79** [jtmoon79/super-speedy-syslog-searcher#1]
tldr Printing _with prepended data_ may insert an errant unicode "unknown glyph" `�`.

### Reproduction

Given file [`./logs/other/tests/dtf9c-12-baddate.log`](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/main/logs/other/tests/dtf9c-12-baddate.log), with sample of contents
```text
$ head -n2 ./logs/other/tests/dtf9c-12-baddate.log
Jan 1 01:00:00 0 😀😁😂😃😄😅😆😇😈😉😊😋😌😍😎😏😐😑😒😓😔😕😖😗😘😙😚😛😜😝😞😟😠😡😢😣😤😥😦😧😨😩😪😫😬😭😮😯😰😱😲😳😴😵😶😷😸😹😺😻😼😽😾😿🙀🙁🙂🙃
Feb 2 02:00:00 1 😁😂😃😄😅😆😇😈😉😊😋😌😍😎😏😐😑😒😓😔😕😖😗😘😙😚😛😜😝😞😟😠😡😢😣😤😥😦😧😨😩😪😫😬😭😮😯😰😱😲😳😴😵😶😷😸😹😺😻😼😽😾😿🙀🙁🙂🙃😀
```

run `s4` as-is (it will print with color, no prepended data).
```text
$ ./target/release/s4 ./logs/other/tests/dtf9c-12-baddate.log | head -n1
Jan 1 01:00:00 0 😀😁😂😃😄😅😆😇😈😉😊😋😌😍😎😏😐😑😒😓😔😕😖😗😘😙😚😛😜😝😞😟😠😡😢😣😤😥😦😧😨😩😪😫😬😭😮😯😰😱😲😳😴😵😶😷😸😹😺😻😼😽😾😿🙀🙁🙂🙃
```

run `s4` with a prepended filename
```text
$ ./target/release/s4 -p ./logs/other/tests/dtf9c-12-baddate.log | head -n1
./logs/other/tests/dtf9c-12-baddate.log:Jan 1 01:00:00 0 😀😁😂😃😄😅😆😇😈😉😊😋😌😍😎😏😐😑😒😓😔😕😖😗😘😙😚😛😜😝😞😟😠😡😢😣😤😥😦😧😨😩😪😫😬😭😮😯😰😱😲😳😴😵😶😷😸😹😺😻�😽😾😿🙀🙁🙂🙃�
```
**ERROR** a two errant `�` is in the output.

This errant output will occur for any mode of prepended data, but at different rates:
- with `-l `  errant `�` occurs 4 times
- with `-p`  errant `�` occurs 20 times
- with `-l -p`  errant `�` occurs 4 times

### Other observations

- Seems to only occur when printing _many_ "large" unicode characters. Does not occur with other test files like [`gen-20-1-☺☻.log`](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/f708e15eab0ca601699461565b7a396f84394526/logs/other/tests/gen-20-1-%E2%98%BA%E2%98%BB.log) or [`./logs/other/tests/gen-20-1-⚀⚁⚂⚃⚄⚅.log`](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/f708e15eab0ca601699461565b7a396f84394526/logs/other/tests/gen-20-1-%E2%9A%80%E2%9A%81%E2%9A%82%E2%9A%83%E2%9A%84%E2%9A%85.log), or [`./logs/other/tests/gen-20-1-🌚🌛🌜🌝.log`](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/f708e15eab0ca601699461565b7a396f84394526/logs/other/tests/gen-20-1-%F0%9F%8C%9A%F0%9F%8C%9B%F0%9F%8C%9C%F0%9F%8C%9D.log).

- the errant `�` does also occurs when printing with or without color; `--color never` or `--color always`.

- In the errant output, the nearby characters are not truncated; the entire file still prints all characters.

- The effect is observed even with blocks larger than the number of bytes in the file (so it's not a "crossing a block" problem).

---

- I inserted debug code into [src/printer/printers.rs](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/f708e15eab0ca601699461565b7a396f84394526/src/printer/printers.rs#L229)
```rust
dpo!("print_color_line_highlight_dt! slice_a.len() {} slice_b_dt.len() {} slice_c.len() {}", slice_a.len(), slice_b_dt.len(), slice_c.len());
```

Running the debug version of the build, with and without `-p` showed the selected slices were the same., e.g.<br/>
without `-p`, debug output
```text
     print_color_line_highlight_dt! slice_a.len() 0 slice_b_dt.len() 15 slice_c.len() 277
Dec 12 12:00:00 11 😋😌😍😎😏😐😑😒😓😔😕😖😗😘😙😚😛😜😝😞😟😠😡😢😣😤😥😦😧😨😩😪😫😬😭😮😯😰😱😲😳😴😵😶😷😸😹😺😻😼😽😾😿🙀🙁🙂🙃😀😁😂😃😄😅😆😇😈😉😊
```
with `-p`, debug output
```text
     print_color_line_highlight_dt! slice_a.len() 0 slice_b_dt.len() 15 slice_c.len() 277
./logs/other/tests/dtf9c-12-baddate.log:Dec 12 12:00:00 11 😋😌😍😎😏😐😑😒😓😔😕😖😗😘😙😚😛😜😝😞😟😠😡😢😣😤😥😦😧😨😩😪😫😬😭😮😯😰😱😲😳😴😵😶😷😸😹😺😻😼😽😾😿🙀🙁🙂🙃😀😁�😃😄😅😆😇😈😉😊�
```

---

- Comparing the first line of output when using `-p`, [`colordiff.pl`](https://github.com/jjazzboss/colorbindiff) shows ***no errant difference***

```text
$ ./target/release/s4 -p ./logs/other/tests/dtf9c-12-baddate.log | head -n1 > out-p1.txt

$ ./target/release/s4 ./logs/other/tests/dtf9c-12-baddate.log | head -n1 > out1.txt

$ perl ./colorbindiff.pl out1.txt out-p1.txt
OFFSET   00  01  02  03  04  05  06  07  08  09  0A  0B  0C  0D  0E  0F                   OFFSET   00  01  02  03  04  05  06  07  08  09  0A  0B  0C  0D  0E  0F
0x0000*  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  .. ................  0x0000* +2e +2f +6c +6f +67 +73 +2f +6f +74 +68 +65 +72 +2f +74 +65 +73 ./logs/other/tes
0x0000*  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  .. ................  0x0010* +74 +73 +2f +64 +74 +66 +39 +63 +2d +31 +32 +2d +62 +61 +64 +64 ts/dtf9c-12-badd
0x0000*  ..  ..  ..  ..  ..  ..  ..  ..                                 ........          0x0020* +61 +74 +65 +2e +6c +6f +67 +3a                                 ate.log:
0x0000   1b  5b  30  6d  1b  5b  33  36  6d  1b  5b  30  6d  1b  5b  34 .[0m.[36m.[0m.[4  0x0028   1b  5b  30  6d  1b  5b  33  36  6d  1b  5b  30  6d  1b  5b  34 .[0m.[36m.[0m.[4
0x0010   6d  1b  5b  33  36  6d  4a  61  6e  20  31  20  30  31  3a  30 m.[36mJan 1 01:0  0x0038   6d  1b  5b  33  36  6d  4a  61  6e  20  31  20  30  31  3a  30 m.[36mJan 1 01:0
0x0020   30  3a  30  30  1b  5b  30  6d  1b  5b  33  36  6d  20  30  20 0:00.[0m.[36m 0   0x0048   30  3a  30  30  1b  5b  30  6d  1b  5b  33  36  6d  20  30  20 0:00.[0m.[36m 0
0x0030   f0  9f  98  80  f0  9f  98  81  f0  9f  98  82  f0  9f  98  83 ................  0x0058   f0  9f  98  80  f0  9f  98  81  f0  9f  98  82  f0  9f  98  83 ................
0x0040   f0  9f  98  84  f0  9f  98  85  f0  9f  98  86  f0  9f  98  87 ................  0x0068   f0  9f  98  84  f0  9f  98  85  f0  9f  98  86  f0  9f  98  87 ................
0x0050   f0  9f  98  88  f0  9f  98  89  f0  9f  98  8a  f0  9f  98  8b ................  0x0078   f0  9f  98  88  f0  9f  98  89  f0  9f  98  8a  f0  9f  98  8b ................
0x0060   f0  9f  98  8c  f0  9f  98  8d  f0  9f  98  8e  f0  9f  98  8f ................  0x0088   f0  9f  98  8c  f0  9f  98  8d  f0  9f  98  8e  f0  9f  98  8f ................
0x0070   f0  9f  98  90  f0  9f  98  91  f0  9f  98  92  f0  9f  98  93 ................  0x0098   f0  9f  98  90  f0  9f  98  91  f0  9f  98  92  f0  9f  98  93 ................
0x0080   f0  9f  98  94  f0  9f  98  95  f0  9f  98  96  f0  9f  98  97 ................  0x00A8   f0  9f  98  94  f0  9f  98  95  f0  9f  98  96  f0  9f  98  97 ................
0x0090   f0  9f  98  98  f0  9f  98  99  f0  9f  98  9a  f0  9f  98  9b ................  0x00B8   f0  9f  98  98  f0  9f  98  99  f0  9f  98  9a  f0  9f  98  9b ................
0x00A0   f0  9f  98  9c  f0  9f  98  9d  f0  9f  98  9e  f0  9f  98  9f ................  0x00C8   f0  9f  98  9c  f0  9f  98  9d  f0  9f  98  9e  f0  9f  98  9f ................
0x00B0   f0  9f  98  a0  f0  9f  98  a1  f0  9f  98  a2  f0  9f  98  a3 ................  0x00D8   f0  9f  98  a0  f0  9f  98  a1  f0  9f  98  a2  f0  9f  98  a3 ................
0x00C0   f0  9f  98  a4  f0  9f  98  a5  f0  9f  98  a6  f0  9f  98  a7 ................  0x00E8   f0  9f  98  a4  f0  9f  98  a5  f0  9f  98  a6  f0  9f  98  a7 ................
0x00D0   f0  9f  98  a8  f0  9f  98  a9  f0  9f  98  aa  f0  9f  98  ab ................  0x00F8   f0  9f  98  a8  f0  9f  98  a9  f0  9f  98  aa  f0  9f  98  ab ................
0x00E0   f0  9f  98  ac  f0  9f  98  ad  f0  9f  98  ae  f0  9f  98  af ................  0x0108   f0  9f  98  ac  f0  9f  98  ad  f0  9f  98  ae  f0  9f  98  af ................
0x00F0   f0  9f  98  b0  f0  9f  98  b1  f0  9f  98  b2  f0  9f  98  b3 ................  0x0118   f0  9f  98  b0  f0  9f  98  b1  f0  9f  98  b2  f0  9f  98  b3 ................
0x0100   f0  9f  98  b4  f0  9f  98  b5  f0  9f  98  b6  f0  9f  98  b7 ................  0x0128   f0  9f  98  b4  f0  9f  98  b5  f0  9f  98  b6  f0  9f  98  b7 ................
0x0110   f0  9f  98  b8  f0  9f  98  b9  f0  9f  98  ba  f0  9f  98  bb ................  0x0138   f0  9f  98  b8  f0  9f  98  b9  f0  9f  98  ba  f0  9f  98  bb ................
0x0120   f0  9f  98  bc  f0  9f  98  bd  f0  9f  98  be  f0  9f  98  bf ................  0x0148   f0  9f  98  bc  f0  9f  98  bd  f0  9f  98  be  f0  9f  98  bf ................
0x0130   f0  9f  99  80  f0  9f  99  81  f0  9f  99  82  f0  9f  99  83 ................  0x0158   f0  9f  99  80  f0  9f  99  81  f0  9f  99  82  f0  9f  99  83 ................
0x0140   0a                                                             .                 0x0168   0a
```
The difference is as-expected: the output with `-p` has extra ASCII bytes to print the prepended filename. However, there are no errant control codes or partial unicode characters.

---

Using Windows Terminal 1.15, Ubuntu 22.
